yt_idv/shaders/ray_tracing.frag.glsl

```diff
@@ -45,6 +50,106 @@ vec4 cleanup_phase(in vec4 curr_color, in vec3 dir, in float t0, in float t1);
 //   void (vec3 tex_curr_pos, inout vec4 curr_color, float tdelta, float t,
 //         vec3 direction);
 
+float get_ray_plane_intersection(vec3 p_normal, float p_constant, vec3 ray_origin, vec3 ray_dir)
+{
+    float n_dot_u = dot(p_normal, ray_dir);
+    float n_dot_ro = dot(p_normal, ray_origin);
+    // add check for n_dot_u == 0 (ray is parallel to plane)
+    if (n_dot_u == float(0.0))
+    {
+        // the ray is parallel to the plane. there are either no intersections
+        // or infinite intersections. In the second case, we are guaranteed
+        // to intersect one of the other faces, so we can drop this plane.
+        return INFINITY;
+    }
+
+    return (p_constant - n_dot_ro) / n_dot_u;
+}
+
+vec2 get_ray_sphere_intersection(float r, vec3 ray_origin, vec3 ray_dir)
+{
+    float dir_dot_dir = dot(ray_dir, ray_dir);
+    float ro_dot_ro = dot(ray_origin, ray_origin);
+    float dir_dot_ro = dot(ray_dir, ray_origin);
+    float rsq = r * r; // could be calculated in vertex shader
+
+    float a_2 = 2.0 * dir_dot_dir;
+    float b = 2.0 * dir_dot_ro;
+    float c =  ro_dot_ro - rsq;
+    float determinate = b*b - 2.0 * a_2 * c;
+    float cutoff_val = 0.0;
+    if (determinate < cutoff_val)
+    {
+        return vec2(INFINITY, INFINITY);
+    }
+    else if (determinate == cutoff_val)
+    {
+        return vec2(-b / a_2, INFINITY);
+    }
+    else
+    {
+        return vec2((-b - sqrt(determinate))/ a_2, (-b + sqrt(determinate))/ a_2);
+    }
+
+}
+
+vec2 get_ray_cone_intersection(float theta, vec3 ray_origin, vec3 ray_dir)
+{
+    // note: cos(theta) and vhat could be calculated in vertex shader
+    float costheta;
+    vec3 vhat;
+    if (theta > PI/2.0)
+    {
+        // if theta is past PI/2, the cone will point in negative z and the
+        // half angle should be measured from the -z axis, not +z.
+        vhat = vec3(0.0, 0.0, -1.0);
+        costheta = cos(PI - theta);
+    }
+    else
+    {
+        vhat = vec3(0.0, 0.0, 1.0);
+        costheta = cos(theta);
+    }
+    float cos2t = pow(costheta, 2);
+    // note: theta = PI/2.0 is well defined. determinate = 0 in that case and
```

Comment:
it may also be worth exploring using slight offsets instead of accounting for singularities; i.e., if `==0.0` produces a singularity, use `0.0 + epsilon`